### ISSUE TITLE
Iterator not fetching next batch records

### ISSUE THREAD

**Issue by AJku7** [deeplearning4j/deeplearning4j#3422]
RecordReaderDatasetIterator not fetching next batch files:
 Only the first batch records are fed to the network all the time.I am trying to input new batches for each epoch.So i am unable to load the complete dataset.

`	final int numInput = 4;
		final int numOutput = 1;
		final int numHidden = 20;
		final double learningRate = 0.03;
		final int batchSize =10;
		final int nEpochs = 10;
		final int seed = 12345;
		final int numSamples=158;

		final File trainFolder = new File("/home/aj/my/samples/B/new.csv");
		
		RecordReader trainReader = null;
		RecordReaderDataSetIterator trainIterator = null;
		DataSet trainData = null;
		MultiLayerConfiguration congifure = null;
		MultiLayerNetwork model = null;
		
		trainReader = new CSVRecordReader(1, ",");
		trainReader.initialize(new FileSplit(trainFolder));
		trainIterator = new RecordReaderDataSetIterator(trainReader,batchSize,4,4,true);
		//trainIterator = new RecordReaderDataSetIterator(trainReader,null,batchSize,4,4,0,15,true);
		
		while(trainIterator.hasNext()){
		trainData = trainIterator.next();
		log.info("Building Model");
		 congifure = new NeuralNetConfiguration.Builder()
				.optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
				.learningRate(learningRate)
				.seed(seed)
				.iterations(1)
				.weightInit(WeightInit.RELU)
				.updater(Updater.RMSPROP)
				.list()
				.layer(0, 
						new DenseLayer.Builder().nIn(numInput).nOut(numHidden)
								.activation(Activation.RELU).build()
						)
				.layer(1,
						new DenseLayer.Builder().nIn(numHidden).nOut(numHidden)
								.activation(Activation.RELU).build()
						)
				
				.layer(2,
						new DenseLayer.Builder().nIn(numHidden).nOut(numHidden)
						.activation(Activation.RELU).build())
				.layer(3,
						new DenseLayer.Builder().nIn(numHidden).nOut(numHidden)
						.activation(Activation.RELU).build())
				.layer(4,
						new OutputLayer.Builder().lossFunction(LossFunctions.LossFunction.MSE)
								.nIn(numHidden).nOut(numOutput)
								.activation(Activation.IDENTITY).build()
						)
				.build();

		log.info("initlizing model");
		model = new MultiLayerNetwork(congifure);
		model.init();
		model.setListeners(new ScoreIterationListener(1));
		
		log.info("Training Model");
		for (int n = 0; n < nEpochs; n++) {
				
			model.fit(trainIterator);
			System.out.println(trainData);
			trainIterator.reset();
			
		}
		}
		
}

}`
The dataset can be found in the dataset.zip file.
[dataset.zip](https://github.com/deeplearning4j/deeplearning4j/files/1007364/dataset.zip)

**Comment by marcel-93** [comment]
imo it's pretty strange what you are doing there. You have a endless loop in which you always config and init a model. i think you should orientate more on the examples and then try to convert it to your data and network structure.

**Comment by agibsonccc** [comment]
Ohhh - yeah that's it. Just create your iterator and pass it in the for loop. I didn't see that. Don't do the while has next in there.

**Comment by AJku7** [comment]
Thanks i tried removing while loop.But still when i print the contents in dataset the same contents is loading every time.
I need to load the complete dataset. So that the iterator can fetch different examples for each epoch.

**Comment by marcel-93** [comment]
https://deeplearning4j.org/troubleshootingneuralnets#epochs
Per definition an epoch is a full pass of the dataset which is done by the model.fit(..) so fetch different data each epoch is strange

**Comment by AJku7** [comment]
@marcel-93 then the program i have mentioned will train the model for complete dataset.And batchsize is for updating weights .
iterator=(batchsize,....)
model.fit(iterator) .here does the model fit only batchsize number of records? or Irrespective of batchsize for complete dataset

**Comment by agibsonccc** [comment]
@AJku7 an iterator *does* an epoch for you. Please get rid of the loop like he's suggesting. I wrote that code. I know what it does.

**Comment by AJku7** [comment]
Thank you both for your help
